### PR TITLE
tsv header pre-processing

### PR DESCRIPTION
**Pull request by jaserud**
**Description of changes**

- refactored clinicalSubmission from api to gateway
- applied pre fieldName check for entities
- updated tests

**some changes are needed in api (and possibly UI?), will open PRs**

**Type of Change**

- [ ] Bug
- [x] Refactor
- [x] New Feature

**Checklist before requesting review:**

- [x] Check branch (code change PRs go to `develop` not master)
- [x] Manual testing
- [x] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)

### CODE REVIEW

File: src/submission/submission-api.ts [hunk 1]
```diff
@@ -131,13 +133,34 @@ class SubmissionController {
       const fileName = filesByTypeMap[clinicalFileType].originalname;
       let records: ReadonlyArray<Readonly<{ [key: string]: string }>> = [];
       try {
-        records = await TsvUtils.tsvToJson(filesByTypeMap[clinicalFileType].path);
+        const clinicalFieldNamesByPriorityMap = schemaManager
```
Comment (blabadi) [inline src/submission/submission-api.ts]: checking the headers should be a domain validation since it's structure related

Comment (blabadi) [reply src/submission/submission-api.ts]: i.e. it should be in submission service.ts

File: src/submission/submission-entities.ts [hunk 2]
```diff
@@ -161,15 +173,29 @@ export interface CreateSubmissionResult {
   readonly submission: DeepReadonly<ActiveClinicalSubmission> | undefined;
   readonly successful: boolean;
   schemaErrors: DeepReadonly<{ [clinicalType: string]: SubmissionValidationError[] }>;
+  fileErrors: DeepReadonly<SubmissionBatchError[]>;
```
Comment (blabadi) [inline src/submission/submission-entities.ts]: batchErrors

File: src/submission/submission-entities.ts [hunk 3]
```diff
@@ -65,6 +65,18 @@ export type SubmissionValidationUpdate = {
   index: number;
 };
 
+export type SubmissionBatchError = {
+  msg: string;
+  fileNames: string[];
```
[the remark this answers is not in the document]
Comment (jaserud) [reply src/submission/submission-entities.ts]: haha fixed it in last commit, it is possible to have multiple batchFiles like for code `MULTIPLE_TYPED_FILES`

File: src/submission/submission-service.ts [hunk 4]
```diff
@@ -24,18 +24,24 @@ import {
...
 import {
   SchemaValidationError,
   TypedDataRecord,
   DataRecord,
   SchemaProcessingResult,
+  SchemaValidationErrorTypes,
 } from '../lectern-client/schema-entities';
 import { loggerFor } from '../logger';
-import { Errors, F } from '../utils';
+import { Errors, F, isStringMatchRegex, TsvUtils } from '../utils';
```
Comment (blabadi) [inline src/submission/submission-service.ts]: I can't find TsvUtils used here?

File: src/submission/submission-service.ts [hunk 5]
```diff
@@ -24,18 +24,24 @@ import {
...
 import { DeepReadonly } from 'deep-freeze';
-import { FileType } from './submission-api';
+import { FileType, FileNameRegex } from './submission-api';
```
Comment (blabadi) [inline src/submission/submission-service.ts]: FileNameRegex, move to the submission-service.ts and rename to BatchNameRegex

Comment (blabadi) [reply src/submission/submission-service.ts]: the service shouldn't use anything or have any reference on the API, FileType should be moved too, and renamed to ClinicalSubmissionEntity or something more domain related

**Pull request merged**